Adds minetest game mapgen configuration parameters

dacmot opened the pull request:
And a few other mapgen goodies. See #60 and the [readme](https://codeberg.org/Dacmot/minetest_mapgen_config/src/branch/main/README.md) for more info.

Lazerbeak12345 commented on mods/mapgen/minetest_mapgen_config/biomes.lua:
```diff
@@ -0,0 +1,199 @@
+            def.node_dungeon_stair = "stairs:stair_snowblock"
+        end
+
+    end
```
Is a fallback for third party biomes needed? (Ex: what if a user added ethereal, or any other biome mod?)

dacmot replied:
Other non-MTG biomes should remain untouched... unless I screwed something up ;)

Support for popular biome mods like etherreal could be added in the future.

Lazerbeak12345 commented on mods/mapgen/minetest_mapgen_config/biomes.lua:
```diff
@@ -0,0 +1,199 @@
+
+
+    -- Keep only the "interesting" biomes in floatlands
+    if (string.find("coniferous_forest deciduous_forest savanna rainforest taiga desert grassland", name)) then
```
Perhaps a blacklist would be better here than a whitelist. I'm sure there's third party biomes that would look good here.

dacmot replied:
Excellent point. I will change that.

Lazerbeak12345 replied:
Alternatively, a plan to have this list be a setting would work very well also. No need to get locked into one particular way of doing things just yet lol.

Lazerbeak12345 commented on mods/mapgen/minetest_mapgen_config/decorations.lua:
```diff
@@ -0,0 +1,28 @@
+-- This file contains modifications to the decorations for map/world generation
+--
+
+local rdecor = table.copy(minetest.registered_decorations)
```
Might be a good idea to comment why. To the untrained eye this looks like it's doing nothing.

dacmot replied:
There's a comment explaining why in `init.lua`. I can copy it again here if you think it worthwhile.

Lazerbeak12345 replied:
I think that might be fine. I'd consider it your decision, as it's your mod.

Lazerbeak12345 commented on mods/mapgen/minetest_mapgen_config/minetest.conf:
```diff
@@ -0,0 +1,55 @@
+# -----------------------
+##############################################################
+# Ores
+mapgen_conf.tin_ground_layer1_scarcity = 1000
```
If I add a third party ores mod, like moreores, would it work without mandatory configuration?

Lazerbeak12345 replied:
Yes

Lazerbeak12345 commented on mods/mapgen/minetest_mapgen_config/ores.lua:
```diff
@@ -0,0 +1,58 @@
+minetest.clear_registered_ores()
+
+for _, def in pairs(rores) do
+    if (def and def.ore) then
```
Lgtm!

Lazerbeak12345 reviewed (approved):
Lgtm!

dacmot commented:
Sorry. Pushed a bit too soon. Still a few issues to iron out...

Lazerbeak12345 reviewed (changes requested):
Sorry, I forgot I had questions and concerns... I need to work on that.

dacmot commented:
No worries!

Lazerbeak12345 commented:
> Sorry. Pushed a bit too soon. Still a few issues to iron out...

Absolutely fine!

dacmot commented:
I've playtested it a fair bit, but it might need some trying from a different perspective. Let me know what you think.

Lazerbeak12345 commented:
Playtest looks good. I didn't notice any significant change, aside from these:

- I couldn't find the cloud-lands.
- For whatever reason I found two dongeons made of `mtg_plus:silver_sandstone_gold` and `mtg_plus:silver_sandstone_cobble`. I'm pretty sure this was already a feature, but I've never seen it before.

dacmot commented:
I realized I never explained everything this PR changes, and many of those changes are not immediately apparent on a quick playtest. So here's a list more or less complete summary:
* Change default cloudlands settings in minetest.conf
  *  Altitude increased to 2000 
    * cloudlands_altitude = 2000
  * I don't remember why the other cloudland settings... I feel like I took them out of the mod's settingtypes.txt but looking at it now, I didn't change them all from the defaults.
* Modify default dirt and sand filler depth, so when you start digging you don't hit rock immediately
  * mapgen_conf.filler_layer_thickness = 20
  * mapgen_conf.ocean_floor_thickness = 5
  * mapgen_conf.riverbed_thickness = 5
* Add a nice mountain-top cold/snow effect above 80
  * mapgen_conf.mountaintops_altitude = 80
  * mapgen_conf.hills_to_mountain_vertical_blend = 30
* Changed the minetest game ore depths, which seem to have changed for the worst (IMHO) recently. Iron can be found at 0, copper -16, tin -32, mese and gold at -64, and diamond at -256. This to me makes no sense because iron is better than copper or bronze, and the digging progression is almost instantly gratified to diamond, the best MTG material. Is this a minecraft thing? /rant
  * mapgen_conf.tin_ground_layer1_ymax = -30
  * mapgen_conf.copper_ground_layer1_ymax = -30
  * mapgen_conf.iron_ground_layer1_ymax = -200
  * mapgen_conf.gold_ground_layer1_ymax = -400
  * mapgen_conf.mese_ground_layer1_ymax = -600
  * mapgen_conf.mese_block_ground_layer1_ymax = -1000
  * mapgen_conf.diamond_ground_layer1_ymax = -1000
* Similarly adjusted the YellowCrystals depth to be more progressive and a bit deeper.
  * yellow_crystals.ore1_ymax = -350
  * yellow_crystals.ore2_ymax = -450
  * yellow_crystals.ore3_ymax = -550
  * yellow_crystals.ore4_ymax = -650
  * @Lazerbeak12345 Because of this, should we remove the meseblocks?
* Changed the dungeon materials based on biome to add more variety, and take advantage of having mtg_plus. See mods/mapgen/minetest_mapgen_config/biomes.lua:111 and below for more info. These can obviously be changed again. Not sure they should be parameterized though.

dacmot commented:
P.S. my method for quick play test is to start a local game in creative mode, do /grant all singleplayer, and then fly fast everywhere (and through ground: dungeons are easy to spot by their blockish shape) or use /teleport to reach cloudlands. I also use amidst to see all the biomes so I can teleport to them faster instead of searching.

Lazerbeak12345 commented:
I'm actually liking it so far, now that I know what to look for.

I think the meseblocks are fine - if we move them to be essentially the lowest altitude (and highest - like in the cloud-lands perhaps) form of mese. I'm thinking the altatude ranges (with some overlap) should include all crystals, ore and meseblock, but each in a range relative to their density.

Also, I noticed there weren't any caves in the cloudlands, but that may have just been the seed I was on. I didn't look long.

Lazerbeak12345 commented:
> P.S. my method for quick play test is to start a local game in creative mode, do /grant all singleplayer, and then fly fast everywhere (and through ground: dungeons are easy to spot by their blockish shape) or use /teleport to reach cloudlands. I also use amidst to see all the biomes so I can teleport to them faster instead of searching.
> 
> Let me know if you know a more efficient way ;)

yeah that's the only way I've figured out. I know theres a mod that lets you teleport to different biomes and such. I use that sometimes if I'm testing biomes as well.

dacmot commented:
I don't recall seeing caves or dungeons in cloud islands. It could make a neat effect sometimes, but I think most islands are too small.

Lazerbeak12345 commented:
> I don't recall seeing caves or dungeons in cloud islands. It could make a neat effect sometimes, but I think most islands are too small.

Right. I only point this out because I've seen them in world-gen without these changes before. Most are too small for sure, but really large ones aren't uncommon either.

dacmot commented:
I did some play testing and I could not find any caverns or dungeons as are found at the world surface or underground. I did however find some secret spider nests with fossilized eggs inside the bigger islands. I tested in both our latest release (2023-03-16) and this branch.

On a separate topic, but related to cloudlands: looking at the mod's code and documentation, it looks like it can use the vines from [plantlife](https://github.com/minetest-whynot/whynot-game/issues/82). Could be a nice addition.

Lazerbeak12345 commented:
It doesn't seem to be very easy to find the larger islands, but here's one with a (very) small cave. (Tested on new world, default everything, v7 worldgen)

![screenshot_20230601_112020](https://github.com/minetest-whynot/whynot-game/assets/22641188/9078df88-0b76-41d8-9f28-051054c1b622)

Are you also looking for examples of caves at the surface? I found many of them in my search for dungeons, but couldn't find any dungeons this time. (in an earlier world I had once, there were about 5 surface level in just one biome, but that's not using this branch)

> plantlife

When I get time I plan on taking your earlier advice and adding any mod that qualifies one at a time. Much faster turnaround that way.

dacmot commented:
I use [amidst](https://github.com/Treer/Amidst-for-Minetest) ([forums](https://forum.minetest.net/viewtopic.php?t=19869)) to find the larger islands.

Ah, well, look at that. I tend to use carpathian instead of v7. Did you use whynot, or just MTG? I'll git it another go with v7.

dacmot commented:
Does this count? :D
![screenshot_20230601_211532](https://github.com/minetest-whynot/whynot-game/assets/23099256/1ecb53de-279f-43b0-a138-912621e642f1)

I suspect it might be because of the deeper dirt layer. The bottom of the islands are mostly rocks, and they seem more likely to have odd grooves and craggs.

dacmot commented:
I'm also having difficulty finding dungeons sometimes. In some places however they are very abundant like here:
![screenshot_20230601_222240](https://github.com/minetest-whynot/whynot-game/assets/23099256/85272675-fe25-4736-8dc8-fcaee7934bcc)

Lazerbeak12345 commented:
> Ah, well, look at that. I tend to use carpathian instead of v7. Did you use whynot, or just MTG? I'll git it another go with v7.

I used WhyNot. On this branch.

dacmot commented:
I've been testing with current whynot and plain MTG+cloudlands in v7 with otherwise default. I can't find much in the way of caverns. I also tried reducing the fill layer depth in my branch but other than the expected thinner dirt layer, the floating islands seem very similar to other branches. The closest to caves I get is  when two islands overlap.

Lazerbeak12345 reviewed (approved):
LTGM. I don't see any actual problems, but if there's something that this isn't doing something that you want it to, then I suppose we still have changes that need to be made.

dacmot commented:
I'm not sure I understand your comment in contrast with the approval. Is it an OK, or more of a "whatever, don't care"?

It's been a while, so I re-read the thread and I'm not seeing any outstanding issues other than possibly* losing caverns in cloudlands.

*I didn't see significant differences

Lazerbeak12345 commented:
Sorry for the confusion.

Like you I'm not sure if the cloudland caverns are broken. I'm assuming they aren't, and even if they are I'm not sure if it's that much of a difference.

It all looks good to me. The extra options these changes will give us and our users will be good.

dacmot commented:
OK. Will merge then.
